text/0071-trailing-commas.md

```diff
@@ -0,0 +1,170 @@
+# Trailing Commas
+
+## Related issues and PRs
+
+- Reference Issues: https://github.com/cedar-policy/cedar/issues/1024
+- Implementation PR(s):
+
+## Timeline
+
+- Started: 2024-06-26 
+
+## Summary
+
+The Cedar grammar (both policies and schemas) should accept trailing
+commas whenever possible.
+
+## Basic example
+
+Currently, this is a parse error:
+```
+permit(principal, action, resource) when {
+    { foo : 3, bar : 2, } == context  
+};
+```
+This RFC would make it not a parse error. 
+It would also change this is several places beyond records.
+
+## Motivation
+
+In general, allowing trailing commas makes editing easier.
+Re-arranging, deleting, and copy-pasting code into/from a comma
+separated grammar element will no longer require adjusting commas.
+
+Example:
+```
+{
+    "managers": ["Dave", "Shelly", "Chris"],
+    "projects" : {
+        "Alpha" : {
+            "zones" : ["A", "B"]
+        },
+        "Beta" : {
+            "zones" : ["D", "F"]
+        }
+    }
+}
+```
+The `Beta` project cannot simply be deleted from this record, and
+likewise, new fields cannot be simply pasted into any level of the
+objects, without adjusting commas.
+
+Many languages allow for trailing commas, including:
+* Rust
+* Python
+* Java,
+* Go
+* JavaScript
+* OCaml
+
+(Frustratingly: not JSON!)
+
+In addition, it will make our schema grammar and policy grammar more
+consistent, as the schema grammar allows trailing commas in many places.
+
+## Detailed design
+
+### Cedar Policy Grammar
+
+The following grammar rules change:
+```
+Scope ::= Principal ',' Action ',' Resource
+```
+becomes:
+```
+Scope ::= Principal ',' Action ',' Resource ','?
+```
+
+```
+EntList ::= Entity {',' Entity}
+```
+becomes:
+```
+EntList ::= Entity {',' | ',' Entity} ','?
+```
+
+```
+ExprList ::= Expr {',' Expr}
+```
+becomes:
+```
+ExprList ::= Expr {',' | ',' Expr} ','?
+```
+
+```
+RecInits ::= (IDENT | STR) ':' Expr {',' (IDENT | STR) ':' Expr}
+```
+becomes:
+```
+RecInits ::= (IDENT | STR) ':' Expr {',' | ',' (IDENT | STR) ':' Expr}
+```
+
+### Cedar Schema Grammar
+The schema grammar already allows trailing commas in several places
+(such as record types and entity lists), but not everywhere:
+
+```
+EntTypes  := Path {',' Path}
+```
+becomes:
+
+```
+EntTypes  := Path {',' | ',' Path}
+```
+
+```
+Names     := Name {',' Name}
+```
+becomes:
+```
+Names     := Name {',' | ',' Name} ','?
+```
+
+```
+Idents    := IDENT {',' IDENT}
+```
+becomes:
+```
+Idents    := IDENT {',' | ',' IDENT}
+```
+
+### Cedar Policy Formatter
+This RFC proposes a change to the style guide, that having a trailing
```

Comment:
I think Rust doesn't do this. I actually prefer this way.

Reply:
![image](https://github.com/cedar-policy/rfcs/assets/5495888/0758e34b-7c63-4550-915f-e32b17254ace)

Reply:
TIL. I still think it's unnecessary.

Reply:
Are you saying that you think the formatter should prefer the other way (ex: `[1,2,3,]` would auto-style to `[1,2,3]`) or that the formatter should not modify what the user writes. (ex: `[1,2,3,]` would auto-style to `[1,2,3,]` and `[1,2,3]` would auto-style to `[1,2,3]`)

Reply:
It looks like `rustfmt` adds a trailing comma when it wants to split something across multiple lines. That's a good behavior IMO.

Adding a trailing comma in `[1,2]` is kind of ugly, but desirable in the record split across multiple lines. Also, for method calls we really don't want to add the trailing comma (in fact, no Cedar methods or functions take more than one argument)

Reply:
Good callout on method calls

Reply:
I meant option C.

Reply:
Chiming in to agree with @john-h-kastner-aws: my preference is to add the trailing comma when split across multiple lines, remove the trailing comma when not split across multiple lines.  But I don't feel strongly, and @aaronjeline's Option C is also fine with me.